charms/unit/harness.py

```diff
@@ -0,0 +1,122 @@
+#
+# harness.py -- Unit Testing Harness for Layer Charms
+#
+# copyright 2016 Canonical Ltd.
+# TODO: License (Apache or LGPL?)
+#
+
+from glob import glob
+import mock
+import unittest
+
+class HarnessError(Exception): pass
+
+class Harness(unittest.TestCase):
+    '''
+    I am a unit testing harness for Juju Layered charms.
+
+    I automatically mock out any modules referenced in the layers
+    directory that do not actually exist in the layer's source. I
+    maintain an internal table of the mock objects, so that you may
+    add behavior to them if necessary for a test.
+
+    I will also setup a mock for the status handlers in hookenv, so
+    that you may set and confirm status in a test.
+
+    '''
+    def __init__(self, *args, **kwargs):
+        super(Harness, self).__init__(*args, **kwargs)
+        self._to_mock = []
+        self._patchers = []
+        self._local_modules = None
+        self.mocks = {}
+
+    @property
+    def last_status(self):
+        '''
+        Helper for mocked out status list.
+
+        Returns the last status set. Raises a HarnessError if this
+        method is called, but the hookenv mock is not active.
+
+        '''
+        try:
+            return self.statuses[-1]
+        except AttributeError:
+            raise HarnessError(
+                'self.statuses does not exist. Did you initialize a '
+                'harness without passing in mock_hookenv_status=True?')
```

Comment:
I can see the argument for making the "you turned off mocking but still tried to fetch a status" error more obvious, but the message would be slightly misleading in that case.  Even so, if you did leave mocking enabled and there was no status set, you'd get an `IndexError` no matter how you structured your assert (and actually, if you used `self.assertEqual(self.last_status[0], 'active')` instead of `self.assertEqual(self.last_status, ('active', 'foo'))` then the `IndexError` would be misleading because you might assume it had to do with your explicit index instead of the one inside the `last_status` property).

Whether we handle that by pre-populating with `(None, None)` or just adding an `if` condition to the property, I do think that `last_status` with no status having been set should return `(None, None)`.  Actually, I do prefer the `if` condition, though, in case you want to do an assert on the full list of statuses instead of just the last one.  E.g.:

```python
self.assertEqual(self.statuses, [
    ('maintenance', 'installing'),
    ('blocked', 'missing relation'),
    ('active', 'ready'),
])
```